src/sms/core/docs_handler.py

```diff
@@ -0,0 +1,221 @@
+# Copyright 2021 - 2024 Universität Tübingen, DKFZ, EMBL, and Universität zu Köln
+# for the German Human Genome-Phenome Archive (GHGA)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains the implementation of hte DocsHandler, providing the business logic for handling documents."""
+
+import json
+import logging
+from re import split
+from typing import Literal, NamedTuple
+
+from sms.config import Config
+from sms.models import Criteria, DocumentType, UpsertionDetails
+from sms.ports.inbound.docs_handler import DocsHandlerPort
+from sms.ports.outbound.docs_dao import DocsDaoPort
+
+
+def log_and_raise_permissions_error(
+    db_name: str, collection: str, operation: Literal["read", "write"]
+):
+    """Log and raise a PermissionError."""
+    rule = f"{db_name}.{collection}:{operation[0]}"
+    error = PermissionError(
+        f"'{operation.title()}' operations not allowed on db '{db_name}',"
+        + f" collection '{collection}'. No rule found that matches '{rule}'",
+    )
+    logging.error(
+        error,
+        extra={"db_name": db_name, "collection": collection, "operation": operation},
+    )
+    raise error
+
+
+class Permission(NamedTuple):
+    """Represents one permission specifying the allowed ops on a given collection."""
+
+    db_name: str
+    collection: str
+    ops: str
+
+
+class Permissions:
+    """A class to parse and interpret configured permissions for database operations."""
+
+    permissions: list[Permission]
+
+    def __init__(self, permissions: list[str]):
+        """Set up the permissions parser. Permissions should already by validated by
+        the config model.
+        """
+        self.permissions = [
+            Permission(*split(r"\.(.*?):", permission)) for permission in permissions
+        ]
+
+    def get_permissions(self, db_name, collection_name) -> str:
+        """List the operations allowed on the specified collection."""
+        for rule in self.permissions:
+            if rule.db_name in ("*", db_name) and rule.collection in (
+                "*",
+                collection_name,
+            ):
+                return rule.ops
+        return ""
+
+    def can_write(self, db_name: str, collection_name: str) -> bool:
+        """Check if WRITE operations are allowed on the specified collection."""
+        permissions = self.get_permissions(db_name, collection_name)
+        return "w" in permissions
```

Comment:
```suggestion
        return "w" in self.get_permissions(db_name, collection_name)
```

Reply:
Fixed in [f71217c](https://github.com/ghga-de/state-management-service/pull/1/commits/f71217c95008826977207d0f7c25f20665896c95)